blockstore/splitstore/splitstore.go

```diff
@@ -458,266 +616,528 @@ func (s *SplitStore) HeadChange(_, apply []*types.TipSet) error {
+}
 
-	// save the warmup epoch
-	s.warmupEpoch = epoch
-	err = s.ds.Put(warmupEpochKey, epochToBytes(epoch))
-	if err != nil {
-		log.Errorf("error saving warmup epoch: %s", err)
+// transactionally protect a reference to an object
+func (s *SplitStore) trackTxnRef(c cid.Cid) {
+	if !s.txnActive {
+		// not compacting
+		return
 	}
 
-	err = s.ds.Put(markSetSizeKey, int64ToBytes(s.markSetSize))
-	if err != nil {
-		log.Errorf("error saving mark set size: %s", err)
+	if isUnitaryObject(c) {
+		return
 	}
-}
 
-// Compaction/GC Algorithm
-func (s *SplitStore) compact(curTs *types.TipSet) {
-	var err error
-	if s.markSetSize == 0 {
-		start := time.Now()
-		log.Info("estimating mark set size")
-		err = s.estimateMarkSetSize(curTs)
+	if s.txnProtect != nil {
+		mark, err := s.txnProtect.Has(c)
 		if err != nil {
-			log.Errorf("error estimating mark set size: %s; aborting compaction", err)
+			log.Warnf("error checking markset: %s", err)
+			// track it anyways
+		} else if mark {
 			return
 		}
-		log.Infow("estimating mark set size done", "took", time.Since(start), "size", s.markSetSize)
-	} else {
-		log.Infow("current mark set size estimate", "size", s.markSetSize)
-	}
-
-	start := time.Now()
-	if s.fullCompaction {
-		err = s.compactFull(curTs)
-	} else {
-		err = s.compactSimple(curTs)
 	}
-	took := time.Since(start).Milliseconds()
-	stats.Record(context.Background(), metrics.SplitstoreCompactionTimeSeconds.M(float64(took)/1e3))
 
-	if err != nil {
-		log.Errorf("COMPACTION ERROR: %s", err)
-	}
+	s.txnRefsMx.Lock()
+	s.txnRefs[c] = struct{}{}
+	s.txnRefsMx.Unlock()
 }
 
-func (s *SplitStore) estimateMarkSetSize(curTs *types.TipSet) error {
-	var count int64
-	err := s.chain.WalkSnapshot(context.Background(), curTs, 1, s.skipOldMsgs, s.skipMsgReceipts,
-		func(cid cid.Cid) error {
-			count++
-			return nil
-		})
-
-	if err != nil {
-		return err
+// transactionally protect a batch of references
+func (s *SplitStore) trackTxnRefMany(cids []cid.Cid) {
+	if !s.txnActive {
+		// not compacting
+		return
 	}
 
-	s.markSetSize = count + count>>2 // overestimate a bit
-	return nil
-}
-
-func (s *SplitStore) compactSimple(curTs *types.TipSet) error {
-	coldEpoch := s.baseEpoch + CompactionCold
-	currentEpoch := curTs.Height()
-	boundaryEpoch := currentEpoch - CompactionBoundary
+	s.txnRefsMx.Lock()
+	defer s.txnRefsMx.Unlock()
 
-	log.Infow("running simple compaction", "currentEpoch", currentEpoch, "baseEpoch", s.baseEpoch, "coldEpoch", coldEpoch, "boundaryEpoch", boundaryEpoch)
+	quiet := false
+	for _, c := range cids {
+		if isUnitaryObject(c) {
+			continue
+		}
 
-	coldSet, err := s.env.Create("cold", s.markSetSize)
-	if err != nil {
-		return xerrors.Errorf("error creating mark set: %w", err)
-	}
-	defer coldSet.Close() //nolint:errcheck
+		if s.txnProtect != nil {
+			mark, err := s.txnProtect.Has(c)
+			if err != nil {
+				if !quiet {
+					quiet = true
+					log.Warnf("error checking markset: %s", err)
+				}
+				continue
```

Comment:
This is still inconsistent with trackTxnRef.

Reply:
in what way? 
Are you concerned about an error not causing the transaction to abort?
Currently the only way the markset errors is if it has been closed, ie the transaction has been aborted.

Reply:
In trackTxnRef, we track the ref even when we error. _Here_, we skip when we error.

Reply:
ah, yes you are right; let me fix it.

Reply:
fixed in https://github.com/filecoin-project/lotus/pull/6728/commits/77854670f02aff7e5d294e11b3e64e89f5c40e5a